app/src/main/java/ch/epfl/sweng/rps/services/MatchmakingService.kt

```diff
@@ -0,0 +1,81 @@
+package ch.epfl.sweng.rps.services
+
+import android.util.Log
+import ch.epfl.sweng.rps.models.Game
+import com.google.firebase.functions.ktx.functions
+import com.google.firebase.ktx.Firebase
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.tasks.await
+
+class MatchmakingService {
+    private val cloudFunctions = CloudFunctions()
+
+    fun queue(gameMode: Game.GameMode): Flow<QueueStatus> = flow {
+        Log.i("MatchmakingService", "Queueing for game mode ${gameMode.toGameModeString()}")
+        emit(QueueStatus.Queued)
+        Log.i("MatchmakingService", "Sending request to cloud function")
+        val gameId = cloudFunctions.queue(gameMode)
+        Log.i("MatchmakingService", "Received game id $gameId")
+
+        val service = ServiceLocator.getInstance().getGameServiceForGame(gameId)
+        Log.i("MatchmakingService", "Got service $service")
+        emit(QueueStatus.Accepted(service))
+    }
+
+    suspend fun acceptInvitation(invitationId: String): FirebaseGameService {
+        val gameId = cloudFunctions.acceptInvitation(invitationId)
+        return ServiceLocator.getInstance().getGameServiceForGame(gameId)
+    }
+
+    suspend fun currentGame(): FirebaseGameService? {
+        val repo = ServiceLocator.getInstance().repository
+        val games = repo.gamesOfUser(repo.getCurrentUid())
+        if (games.isEmpty()) {
+            return null
+        }
+        if (games.size > 1) {
+            Log.e(
+                "MatchmakingService",
+                "More than one game for user ${repo.getCurrentUid()}: ${games.map { it.id }}"
+            )
+        }
+        return ServiceLocator.getInstance().getGameServiceForGame(games[0].id)
```

Comment:
Hi, could you elaborate on when this class would be used? And how do we know that the game index zero would the current one?

Reply:
Theoretically there should be only one current game played at a time. If there are more, we return one of them, the first. Once this game will be done, you can play the next one.